File: day01/solution.ts
const DEBUG = false;

const data = await Deno.readTextFile('input.txt');
if (DEBUG) console.debug(data);
const lists = data
  .split('\n')
  .filter((line) => line.trim() != '')
  .reduce<{ left: number[]; right: number[] }>(
    (acc, line) => {
      const [leftValue, rightValue] = line.split(/\s+/).map((i) => Number(i));
      if (DEBUG) console.debug({ leftValue, rightValue });
      acc.left.push(leftValue);
      acc.right.push(rightValue);
      return acc;
    },
    { left: [], right: [] }
  );
if (DEBUG) console.debug(lists);

lists.left.sort();
lists.right.sort();

const diffs = Array.from(lists.left, (val, i) => Math.abs(val - lists.right[i]));
if (DEBUG) console.debug(diffs);
const total = diffs.reduce((acc, item) => {
  if (DEBUG) console.debug({ acc, item });
  return acc + item;
}, 0);
console.log(`part 1 ${total}`);

const counts = lists.right.reduce<Record<number, number>>((acc, item) => {
  const prev = acc[item];
  acc[item] = prev ? prev + 1 : 1;
  return acc;
}, {});
if (DEBUG) console.debug(counts);
const similarity_total = lists.left.reduce((acc, item) => {
  acc += item * (counts[item] ?? 0);
  return acc;
}, 0);
console.log(`part 2 ${similarity_total}`);

File: day01/solution.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"slices"
	"strconv"
)

func main(){
	left,right:=readData("input.txt")
	part1(left,right)
	part2(left,right)
}

func part2(left[]int,right[]int){
	rightCounts:=make(map[int]int)
	for _,rightValue:=range right{
		rightCounts[rightValue]++
	}
	simTotal:=0
	for _,leftValue:=range left{
		simTotal+=leftValue*rightCounts[leftValue]
	}
	fmt.Printf("part 2 %d\n",simTotal)
}

func part1(left[]int,right[]int){
	total:=0
	for ix:=range left{
		total+=int(math.Abs(float64(left[ix]-right[ix])))
	}
	fmt.Printf("part 1 %d\n",total)
}

func readData(filename string)(left []int, right []int){
	f,_:=os.Open(filename)
	defer f.Close()
	scanner:=bufio.NewScanner(f)
	scanner.Split(bufio.ScanWords)
	ix:=0;
	for scanner.Scan(){
		parsed,_:=strconv.Atoi(scanner.Text())
		if (ix%2==0){
			left=append(left,parsed)
			} else {
			right=append(right,parsed)
		}
		ix++
	}
	slices.Sort(left)
	slices.Sort(right)
	return
}